### PR TITLE
Arm sim

### PR DESCRIPTION
**Pull request by AnotherUser12**
really cool wrist and arm sim

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/Wrist.java [hunk 1]
```diff
@@ -62,33 +110,18 @@ public void periodic() {
...
   public void updateGoal(){    // check what range the arm is in and set the wrist accordingly
-    double armWrappedPos = arm.getWrappedPosition();
+    double armPos = arm.getPosition();
     for(WristState wristState : WristState.values()){
-      if(wristState.inRange(armWrappedPos)){
-        double goal = wristState.getWristGoal(armWrappedPos);
+      if(wristState.inRange(armPos)){
+        double goal = wristState.getWristGoal(armPos);
         setGoal(goal);
         break;
```
Comment (person4268) [inline src/main/java/frc/robot/subsystems/Wrist.java:126]: should be `return;`, `break;` is for loops

File: src/main/java/frc/robot/subsystems/WristState.java [hunk 2]
```diff
@@ -19,7 +22,7 @@ public boolean inRange(double currentWrappedPos){
         return currentWrappedPos>=min && currentWrappedPos<=max;
     }
     public double getWristGoal(double currentWrappedPos){
-        return this.getWristGoalLambda.apply(currentWrappedPos);
+        return (this.getWristGoalLambda.apply(currentWrappedPos));
```
Comment (person4268) [inline src/main/java/frc/robot/subsystems/WristState.java:25]: nit: unnecessary parentheses

File: src/main/java/frc/robot/subsystems/Wrist.java [hunk 3]
```diff
@@ -62,33 +110,18 @@ public void periodic() {
...
     // This method will be called once per scheduler run
-    double currentPose = getWrappedPosition();
-    double closerGoal = getClosestGoal(goal);
-    double pidval = pid.calculate(currentPose, closerGoal);
-    double ffval = ff.calculate(currentPose+Math.PI/2, 0);
+    double currentPose = getPosition();
+    double pidval = pid.calculate(currentPose, goal);
+    double ffval = ff.calculate(currentPose, 0);
     motor.setVoltage(pidval + ffval);
+    this.simVolts=pidval+ffval;
```
Comment (person4268) [inline src/main/java/frc/robot/subsystems/Wrist.java:117]: can you factor out all instances of stuff like this into its own private setVoltage function? just like you have have getPosition

File: src/main/java/frc/robot/controllers/OperatorController.java [hunk 4]
```diff
@@ -1,7 +1,32 @@
...
+  public default Trigger armGoToInsideRobotNode() {
+    return new Trigger(()->false);
+  }
+  public default Trigger armGoToGroundPickUpNode() {
+    return new Trigger(()->false);
+  }
+  public default Trigger setArmByJoystick() {
+    return new Trigger(()->{
+      return false;
+    });
```
Comment (person4268) [inline src/main/java/frc/robot/controllers/OperatorController.java:30]: minor: should probably make this match all the other ones

File: src/main/java/frc/robot/controllers/Controllers.java [hunk 5]
```diff
@@ -39,7 +39,8 @@ public static void updateActiveControllerInstance() {
     for (int i = 0; i < DriverStation.kJoystickPorts; i++) {
       String joyName = DriverStation.getJoystickName(i);
       if (joyName.equals("")) continue;
-      if(!foundOperatorController && joyName.toLowerCase().contains("logitech")) {
+      if(!foundOperatorController && (joyName.toLowerCase().contains("logitech") ||
+                                      joyName.toLowerCase().contains("keyboard") )) {
```
Comment (person4268) [inline src/main/java/frc/robot/controllers/Controllers.java:43]: doesn't this make it impossible to drive a sim robot with a keyboard? you'd want that especially for vision sim

File: src/main/java/frc/robot/subsystems/Arm.java [hunk 6]
```diff
@@ -62,22 +134,38 @@ private void configure() {
...
+        double ffval = ff.calculate(pose, 0); // ff wants 0 parallel to floor in pos x
+        setVoltage(ffval+pidval);
       }
         break;
-      case VEL:
-      {
-        double ffval = ff.calculate(wrappedPose+Math.PI/2, goal);
-        motor.setVoltage(ffval);
+      case VEL: {
+        if((getPosition()>Units.degreesToRadians(0) && goal>0) || (getPosition()<-Units.degreesToRadians(180) && goal<0)){
```
Comment (person4268) [inline src/main/java/frc/robot/subsystems/Arm.java:162]: ideally you'd comment why such checks are necessary or at least what they do

File: src/main/java/frc/robot/subsystems/Arm.java [hunk 7]
```diff
@@ -9,28 +9,97 @@
...
+  public void simulationPeriodic() {
+    // In this method, we update our simulation of what our arm is doing
+    // First, we set our "inputs" (voltages)
+    m_armSim.setInput(this.simVolts);
 
-  public enum ArmControlMode{
+    // Next, we update it. The standard loop time is 20ms.
+    m_armSim.update(0.020);
+
+    // SimBattery estimates loaded battery voltages
```
Comment (person4268) [inline src/main/java/frc/robot/subsystems/Arm.java:71]: i don't see a SimBattery here

File: src/main/java/frc/robot/subsystems/Arm.java [hunk 8]
```diff
@@ -9,28 +9,97 @@
...
+
   public ArmControlMode controlMode;
-  
+  private double simVolts = 0;
+
   public Arm() {
-    controlMode=ArmControlMode.VEL;
+    controlMode = ArmControlMode.VEL;
     configure();
+    SmartDashboard.putData("Arm Sim", m_mech2d);
```
Comment (person4268) [inline src/main/java/frc/robot/subsystems/Arm.java:87]: Consider naming this just "Arm", considering that publishing arm state is useful outside of sim

File: src/main/java/frc/robot/RobotContainer.java [hunk 9]
```diff
@@ -99,7 +107,10 @@ public RobotContainer() {
     
     vision = new Vision(swerve);
     autonChooser = new AutonChooser(swerve);
-    // Configure the button bindings
+    arm = new Arm();
+    this.wrist=new Wrist(arm);
```
Comment (person4268) [inline src/main/java/frc/robot/RobotContainer.java:111]: minor: for consistency, just do `wrist = new Wrist()`, no this

**Review by person4268 — CHANGES_REQUESTED** [review]
Some other things to consider doing:
1. Refactor sim code into an ArmSim class
2. Annotate functions, if you'd like to spare the time to
3. Create a formatting commit, just right click and hit format document on all the files you've touched
4. Comment any checks that aren't immediately obvious why they're there

File: src/main/java/frc/robot/controllers/LogitechOperatorController.java [hunk 10]
```diff
@@ -36,7 +36,12 @@ public Trigger armGoToGroundPickUpNode() {
   }
   public Trigger setArmByJoystick() {
     return new Trigger(()->{
-      return Math.abs(getArmAxis())>0.2;
+      if(Math.abs(getArmAxis())>0.2){
+        if(Math.random()>0.5){
+          return true;
+        }
+      }
```
Comment (person4268) [inline src/main/java/frc/robot/controllers/LogitechOperatorController.java:43]: what even

File: src/main/java/frc/robot/controllers/LogitechOperatorController.java [hunk 11]
```diff
@@ -36,7 +36,12 @@ public Trigger armGoToGroundPickUpNode() {
   }
   public Trigger setArmByJoystick() {
     return new Trigger(()->{
-      return Math.abs(getArmAxis())>0.2;
```
Comment (person4268) [inline src/main/java/frc/robot/controllers/LogitechOperatorController.java]: also like just do Controllers.deadband(), you don't need this at all (unless you needed a different deadband, in which case just modify deadband to have a customizable one)

File: src/main/java/frc/robot/subsystems/MechRootWrapper.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+    public MechanismLigament2d append(MechanismLigament2d lig){
+        this.root.append(lig);
+        return lig;
+    }
+
+    public Mechanism2d getMechanism2d(){
+        return mech;
+    }
+}
```
Comment (person4268) [inline src/main/java/frc/robot/subsystems/MechRootWrapper.java:40]: given how many classes you have, it's probably a good idea to put them all in their own subdirectory

File: src/main/java/frc/robot/controllers/Controllers.java (hunk 5 above)
Comment (AnotherUser12) [reply src/main/java/frc/robot/controllers/Controllers.java:43]: prolly unnecessary

**Comment by AnotherUser12** [comment]
probably works: merging - michael

**Pull request merged**